src/relay/ir/module.cc

```diff
@@ -110,15 +111,22 @@ void ModuleNode::Add(const GlobalVar& var,
   AddUnchecked(var, checked_func);
 }
 
+void ModuleNode::RegisterConstructors(const GlobalTypeVar& var, const TypeData& type) {
```

Comment:
Using hash value as tag makes opcode not very intuitive to understand, How about maintaining a continuous range [0, num_constructors) for each ADT type? 

Then for the following code:
````ocaml
type t = | Alice | Bob | Charlie | David

let test v =
  match v with
  | Alice   -> 1
  | Bob     -> 2
  | Charlie -> 3
  | David   -> 4
````

We can generate opcode like:
````
(switch* v
          case int 0: 1
          case int 1: 2
          case int 2: 3
          case int 3: 4)
````

Reply:
That's a good point and I agree hashes are not ideal. The reason I went with hashes is to ensure that the tag will be determined completely by the name of the global type var and the index of the constructor in the type data and thus will not be affected, for example, by the order in which types are defined. It also means that tags will not collide across different ADTs. If hashes are undesirable, we can take the approach (as I alluded to before) of requiring a type check and only looking at the index of the constructor.

Reply:
Yeah, I vote for eliminating the hash.

Reply:
Should the constructor value carry information about its type, then? The case that is most difficult is the one in `arg_to_ast`: given only the constructor value, how do you work backwards?

Reply:
I think for VM, we might need to discard type information in `Object`. As we already drop the type information in byte codes we generate.

Reply:
I'll see what I can do about `arg_to_ast` then

Reply:
I'm not sure, though, that it will be possible to drop type information for ADT values without ensuring that all tags be globally unique (and I'm not sure of any better way to do that than hashing).

Reply:
Yeah I agree there is no other way . And seems like interpreter has the need to map runtime value back to AST node to support user provided bindings. 

How about make a 8 bit hash instead and put it in the most significant byte of the tag. So when we debug, we can easily tell the variant from the least significant byte.  What do you think?

Reply:
Interesting suggestion, definitely more readable too. I will look into implementing that, thanks.